### PR TITLE
feat(doctor): tool update advisories with per-tool update-channel resolver

### PR DESCRIPTION
**Pull request by TurboCoder13**
## Commit Summary (Conventional Commits)

- Title (required, present tense):

  ```text
  feat(doctor): tool update advisories with per-tool update-channel resolver
  ```

- Type:
  - [x] feat (minor)
  - [ ] fix / perf (patch)
  - [ ] docs
  - [ ] refactor
  - [ ] test
  - [ ] chore / ci / style

- Breaking change:
  - [ ] `!` in title or `BREAKING CHANGE:` footer included

### Release Trigger Rules (exact)

- A merged PR will bump the version based on its title (squash merge required):
  - `feat(...)` or `feat:` → MINOR bump
  - `fix(...)` / `fix:` or `perf(...)` / `perf:` → PATCH bump
  - Any title with `!` after the type (e.g. `feat!:` or `feat(scope)!:`) or a body
    containing `BREAKING CHANGE:` → MAJOR bump
- Use squash merge so the PR title becomes the merge commit title.

## What’s Changing

Teach `lintro doctor` / `lintro versions` to surface actionable "update available"
advisories per tool by resolving how each binary was installed (Homebrew, uv tool,
pip/venv, npm/bun, cargo, rustup, standalone) and mapping that channel to an update
command. Latest-known versions come from pinned manifest / tool-versions data — no
network calls in v1.

- New `UpdateChannel` enum + `update_channels.py` path heuristics
- `VersionAdvisory` produced from `version_checking.build_version_advisory`
- Doctor / versions render an Update advisories section and expose structured JSON
- Optional manifest `install.update_channel` override when heuristics fail
- Unknown / standalone channels degrade gracefully (advisory without inventing a command)

Does **not** implement #1244 (cached capability probing).

## Checklist

- [x] Title follows Conventional Commits
- [x] Tests added/updated
- [ ] Docs updated if user-facing
- [x] Local CI passed (`uv run lintro fmt/chk`, `uv run lintro tst` / full pytest)

## Closes

- Closes #1243

## Details

Channel detection resolves symlinks so Homebrew Cellar installs linked from
`/opt/homebrew/bin` or `/usr/local/bin` classify correctly (including brew formulas
that nest under `node_modules`). When path detection yields unknown/standalone,
doctor falls back to the manifest `install.type` for a usable update command.

Testing:
- Unit tests with fake paths per channel (no real binaries required)
- Doctor unit coverage for path-based advisory enrichment
- Full `uv run lintro chk` (0 issues) and full pytest (6715 passed)

<details><summary><h3>Greptile Summary</h3></summary>

This PR teaches `lintro doctor` and `lintro versions` to surface per-tool "update available" advisories by detecting how each binary was installed (Homebrew, uv tool, pip/venv, npm/bun, cargo, rustup, standalone) and mapping that to a channel-specific update command template — all without network calls.

- Adds `UpdateChannel` enum, `update_channels.py` with path-heuristic detection and `VersionAdvisory` dataclass, and wires the advisory into `ToolCheckResult` / `ToolVersionInfo` for both doctor and versions output paths.
- Structured advisory is exposed in JSON output; human-readable output shows the detected channel diagnostically alongside the existing install-strategy `upgrade_hint`.
- All three previously-flagged false-positive heuristics (bare `cellar`, broad `toolchains`, `BUN_INSTALL` substring) are resolved, and the import consolidation is done.
</details>

<h3>Confidence Score: 5/5</h3>

- Safe to merge. All changes are additive and advisory-only; they enrich display/JSON output without modifying version-check logic, install commands, or tool execution paths.
- The channel-detection heuristics are isolated, well-tested, and gracefully degrade to UNKNOWN/STANDALONE rather than producing wrong commands. The two quality nits — misleading binary_path in JSON for wrapper-probed tools, and duplicated advisory text between the inline line and the summary block — have no impact on correctness or user safety. ManifestRegistry.load() is lru_cache'd so the new registry lookup inside check_tool_version is not a performance concern.
- No files require special attention. The two comments are in version_parsing.py (binary_path field accuracy) and doctor.py (duplicate advisory rendering), both cosmetic/quality issues.

<details><summary><h3>Important Files Changed</h3></summary>

| Filename | Overview |
|----------|----------|
| lintro/tools/core/update_channels.py | New module implementing path-based install channel detection and update command resolution. Previous thread issues (BUN_INSTALL substring, bare `cellar` match, broad rustup heuristic) are all fixed. Logic is well-structured with clear separation of heuristics per channel. |
| lintro/tools/core/version_checking.py | Adds `build_version_advisory` wrapper with dual-fallback logic (no-path → install_type and UNKNOWN/STANDALONE path → install_type). ManifestRegistry.load() is lru_cache'd so multi-tool runs don't incur repeated I/O. Import consolidation noted in previous thread is done. |
| lintro/tools/core/version_parsing.py | Adds `binary_path` and `advisory` fields to ToolVersionInfo. `binary_path` is set at creation time from `shutil.which(command[0])`, which for wrapper-probed tools (cargo, bash) stores the wrapper path; this is what gets exported in JSON output, not the actual tool binary. |
| lintro/cli_utils/commands/doctor.py | Renders advisories inline per tool and in a new "Update advisories" summary block; both call the same format_advisory_line, producing identical text twice for every outdated/incompatible tool. JSON output now includes structured advisory objects. |
| lintro/cli_utils/commands/versions.py | Adds advisory section, below_recommended status display, and richer JSON output (recommended_version, below_recommended, binary_path, advisory). Clean and well-tested. |
| lintro/utils/doctor_report.py | Adds advisory enrichment to ToolCheckResult for OUTDATED/INCOMPATIBLE statuses. Uses resolve_channel_binary_path correctly to handle wrapper probes. Advisory field is nullable and gracefully absent for non-outdated statuses. |
| tests/unit/tools/core/test_update_channels.py | Comprehensive parametrized coverage of channel detection paths including edge cases: wine-cellar false positive, project toolchains false positive, rustup shims, env-var overrides. All previous regression scenarios from review threads are explicitly tested. |
| tests/unit/cli_utils/commands/test_doctor_command.py | New advisory tests cover path-based detection, manifest overrides, rustup shim handling, wrapper probes (cargo, bash), and rendering. Good coverage of the new behaviour. |

</details>

<details open><summary><h3>Flowchart</h3></summary>

```mermaid
%%{init: {'theme': 'neutral'}}%%
flowchart TD
    A["check_tool / check_tool_version"] --> B["resolve_channel_binary_path\nskip wrapper argv0 like cargo/bash\nfind actual tool binary via which"]
    B --> C["detect_update_channel(binary_path)"]
    C --> D{Override?}
    D -- "channel_override or TOOL_CHANNEL_OVERRIDES" --> E[Return override channel]
    D -- No --> F{Path heuristics}
    F --> G[HOMEBREW]
    F --> H[UV_TOOL]
    F --> I["cargo bin path"]
    I --> J{rustup shim name?}
    J -- Yes --> K[RUSTUP]
    J -- No --> L[CARGO]
    F --> M[rustup toolchains]
    F --> N[bun install]
    F --> O[node_modules]
    F --> P["site-packages / venv"]
    F --> Q[system bin prefix]
    F --> R[UNKNOWN]
    M --> K
    N --> BUN[BUN]
    O --> NPM[NPM]
    P --> PIP[PIP]
    Q --> STANDALONE[STANDALONE]
    E & K & L & BUN & NPM & PIP & STANDALONE & R --> S["resolve_update_command\nbrew upgrade / uv tool upgrade\nnpm install / cargo install / ..."]
    S --> T["VersionAdvisory\ntool, installed, latest_known, channel, update_command"]
    T --> U["JSON: advisory.to_dict"]
    T --> V["Human: format_advisory_line\nchannel label only no command"]
```
</details>

<sub>Reviews (5): Last reviewed commit: ["fix(doctor): classify wrapper probes wit..."](https://github.com/lgtm-hq/py-lintro/commit/924df3164ae35eba24fc8ab726af434ccd153691) | [Re-trigger Greptile](https://app.greptile.com/api/retrigger?id=45056121)</sub>

## Summary by CodeRabbit

* **New Features**
  * Added update advisories to tool diagnostics, version checks, and introspection results.
  * Advisories show installed and recommended versions, installation channels, binary paths, and available update commands.
  * Added detection for Homebrew, uv, pip, npm, Bun, Cargo, rustup, standalone, and unknown installations.
  * Enhanced terminal, Markdown, and JSON output for outdated or incompatible tools.
  * Version summaries now distinguish passing tools from those below recommended versions.

* **Documentation**
  * Updated JSON and introspection documentation with advisory, binary path, and upgrade hint details.

### CODE REVIEW

File: lintro/tools/core/update_channels.py [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+
+
+def _is_bun_path(*, path_lower: str, parts_lower: set[str]) -> bool:
+    """Return True when the path is under a Bun install prefix."""
+    if ".bun" in parts_lower:
+        return True
+    bun_install = os.environ.get("BUN_INSTALL")
+    if bun_install:
+        return bun_install.rstrip("/").lower() in path_lower
+    return "/.bun/" in path_lower
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply lintro/tools/core/update_channels.py:358]: Fixed in effcfb36 — `BUN_INSTALL` (and `CARGO_HOME`) now use `Path.is_relative_to` like `UV_TOOL_DIR`.

File: lintro/tools/core/update_channels.py [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+    """Return True when the path is under a Homebrew prefix."""
+    if "cellar" in parts_lower or "linuxbrew" in parts_lower:
+        return True
+    markers = (
+        "/opt/homebrew/",
+        "/home/linuxbrew/",
+        "/usr/local/homebrew/",
+        "/homebrew/",
+    )
+    return any(marker in path_lower for marker in markers)
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply lintro/tools/core/update_channels.py]: Fixed in effcfb36 — dropped bare `cellar` matching; Homebrew detection now uses explicit prefixes including `/usr/local/Cellar/` and `/opt/homebrew/Cellar/`.

File: lintro/tools/core/update_channels.py [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+    return False
+
+
+def _is_rustup_path(*, path_lower: str, parts_lower: set[str]) -> bool:
+    """Return True when the path is under a rustup toolchain tree."""
+    if "rustup" in parts_lower:
+        return True
+    return "/toolchains/" in path_lower and (
+        "rust" in path_lower or "clippy" in path_lower or "rustc" in path_lower
+    )
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply lintro/tools/core/update_channels.py]: Fixed in effcfb36 — rustup detection now requires a `rustup` path component or `/.rustup/toolchains/`.

File: lintro/tools/core/version_checking.py [hunk 4]
```diff
@@ -50,14 +50,24 @@
...
 )
 from lintro.enums.tool_name import ToolName
+from lintro.enums.update_channel import UpdateChannel
+from lintro.tools.core.update_channels import (
+    VersionAdvisory,
+    channel_from_install_type,
+)
+from lintro.tools.core.update_channels import (
+    build_version_advisory as _build_version_advisory,
+)
```
[the remark this answers is not in the document]
Comment (TurboCoder13) [reply lintro/tools/core/version_checking.py]: Fixed in effcfb36 — consolidated by importing the module as `update_channel_ops` (ruff/isort splits `import ... as` aliases into a second block, so a module import keeps a single clean call site).
